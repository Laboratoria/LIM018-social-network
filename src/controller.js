import { components } from './views/components.js';

export const changeView = (route) => {
  const containerBody = document.querySelector('body');
  containerBody.innerHTML = '';
  switch (route) {
    case '': {
      components.loginView();
      components.logInBehavior();
      break; }
    case '#/sign-up': { containerBody.appendChild(components.signUpView());
      components.signUpBehavior();
      break;
    }
    case '#/log-in': { components.loginView();
      components.logInBehavior();
      break;
    }
    case '#/home': { containerBody.appendChild(components.homeView());
      components.homeBehavior();
      break;
    }
    default: { containerBody.appendChild(components.errorView());
      break; }
  }
};
