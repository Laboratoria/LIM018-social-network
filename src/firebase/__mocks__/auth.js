export const signUpWithEmail = jest.fn();
export const logInWithEmail = jest.fn();
export const signUpWithGmail = jest.fn();
export const emailVerification = jest.fn();
export const getCurrentUser = jest.fn();
export const logOut = jest.fn();
