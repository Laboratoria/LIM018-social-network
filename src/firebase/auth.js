import {
  getAuth,
  createUserWithEmailAndPassword,
  signInWithEmailAndPassword,
  GoogleAuthProvider,
  signInWithPopup,
  sendEmailVerification,
  updateProfile,
  signOut,
  // eslint-disable-next-line import/no-unresolved
} from 'https://www.gstatic.com/firebasejs/9.6.10/firebase-auth.js';
import { app } from './firebaseConfig.js';

export const auth = getAuth(app);
const provider = new GoogleAuthProvider();

// eslint-disable-next-line max-len
export const signUpWithEmail = (email, password) => createUserWithEmailAndPassword(auth, email, password);

// eslint-disable-next-line max-len
export const logInWithEmail = (email, password) => signInWithEmailAndPassword(auth, email, password);

export const signUpWithGmail = () => signInWithPopup(auth, provider);

export const emailVerification = () => sendEmailVerification(auth.currentUser);

export const getCurrentUser = () => auth.currentUser;

// eslint-disable-next-line max-len
export const completeUserInfo = (userInformation) => updateProfile(auth.currentUser, userInformation);

export const logOut = () => {
  signOut(auth);
};

export { GoogleAuthProvider };
