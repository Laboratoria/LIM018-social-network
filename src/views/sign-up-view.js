// eslint-disable-next-line import/no-unresolved
import {
  signUpWithEmail, signUpWithGmail, emailVerification, completeUserInfo, GoogleAuthProvider,
} from '../firebase/auth.js';
import { userInfoFirestore } from '../firebase/post.js';

export const createSignUpView = () => {
  const viewSignup = `
  <div class='containerForm'>
    <div class='view'>
      <h1 class='titles'> THE SOCIAL FOOD </h1>
        <form>
          <input type='text' placeholder='Nombre de Usuario' id='userName' class="registerInputs">
          <br>
          <input type='email' placeholder='Correo electrónico' id='email' value='' class="registerInputs">
          <br>
          <input type='password' placeholder='Contraseña' id='userPassword' value='' class="registerInputs">
          <br>
          <button type='button' id='submitSingUp' class='buttonsForm' value='Sign Up'>Sign Up</button>
        </form>
        <div id='eMessage'></div>
        <img src="./images/google-logo.png" id="gmailSignIn" class='googleImage'>
        <p>Con Google</p>
        <h3>¿Tienes cuenta?<span><a href='#/log-in'>  Entrar</a></span></h3>
    </div>
    <div class='modalContainer'>
      <div class='modal'>
        <img src='./images/verification-email.jpg'>
        <p>Se envió un código de verificación a su correo electrónico</p>
        <button class='modalButton'>Aceptar</button>
      </div>
    </div>    
  </div>`;

  const sectionElement = document.createElement('section');
  sectionElement.setAttribute('class', 'backgroundImage');
  sectionElement.innerHTML = viewSignup;

  return sectionElement;
};

export const createBehaviorSignUpView = () => {
  const userName = document.querySelector('#userName');
  const userEmail = document.querySelector('#email');
  const userPassword = document.querySelector('#userPassword');
  const eMessage = document.querySelector('#eMessage');
  const submitButton = document.querySelector('#submitSingUp');
  const gmailButton = document.querySelector('#gmailSignIn');
  const modalContainer = document.querySelector('.modalContainer');
  const closeModal = document.querySelector('.modalButton');

  submitButton.addEventListener('click', () => {
    signUpWithEmail(userEmail.value, userPassword.value)
      .then((userCredential) => {
        emailVerification().then(() => {
          modalContainer.classList.add('reveilModal');
        });
        const uName = userName.value;
        const user = userCredential.user;
        completeUserInfo({ displayName: `${uName}` });
        userInfoFirestore(user.uid, {
          email: user.email,
          name: uName,
          uid: user.uid,
        });
      }).catch((error) => {
        console.log(error);
        const errorM = error.message;
        eMessage.setAttribute('class', 'errorMessage');
        // eslint-disable-next-line no-console
        console.log(errorM);
        switch (errorM) {
          case 'Firebase: Error (auth/invalid-email).': {
            eMessage.textContent = 'Debe ingresar un correo electrónico válido';
            break;
          }
          case 'Firebase: Password should be at least 6 characters (auth/weak-password).': {
            eMessage.textContent = 'La contraseña debe tener como mínimo 6 caracteres';
            break;
          }
          case 'Firebase: Error (auth/email-already-in-use).': {
            eMessage.textContent = 'El correo electrónico ya está siendo usado';
            break;
          }
          default: eMessage.textContent = '';
            break;
        }
      });

    closeModal.addEventListener('click', () => {
      modalContainer.classList.remove('reveilModal');
      window.location.href = '#/log-in';
    });
  });

  gmailButton.addEventListener('click', () => {
    signUpWithGmail().then((result) => {
      // This gives you a Google Access Token. You can use it to access the Google API.
      // The signed-in user info.
      sessionStorage.clear();
      GoogleAuthProvider.credentialFromResult(result);
      const user = result.user;
      sessionStorage.setItem('USER', JSON.stringify(user.uid));
      userInfoFirestore(user.uid, {
        email: user.email,
        name: user.displayName,
        photo: user.photoURL,
      });
      // eslint-disable-next-line no-console
      console.log(user);
      // redireccionar y ruteo
      window.location.href = '#/home';
    }).catch((error) => {
      const errorMessage = error.message;
      // The email of the user's account used.
      const email = error.customData.email;
      // eslint-disable-next-line no-console
      console.log(errorMessage, email);
      // The AuthCredential type that was used.
      // const credential = GoogleAuthProvider.credentialFromError(error);
      // ...
    });
  });
};
