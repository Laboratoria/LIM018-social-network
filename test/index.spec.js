// importamos la funcion que vamos a testear
import { changeView } from '../src/controller.js';
import { createSignUpView, createBehaviorSignUpView } from '../src/views/sign-up-view.js';
import { signUpWithEmail, emailVerification } from '../src/firebase/auth.js';
// import { userInfo } from '../src/firebase/post.js';

jest.mock('../src/firebase/auth.js');
jest.mock('../src/firebase/post.js');

describe('changeView', () => {
  document.body.innerHTML = '';
  window.location.href = '#/home';
  it('debería ser una función', () => {
    expect(typeof changeView).toBe('function');
  });

  it('debería imprimir la ruta log-in', () => {
    changeView('#/log-in');
    const body = document.querySelector('body');
    expect(body).toMatchSnapshot();
  });

  it('debería imprimir la ruta sign-up', () => {
    changeView('#/sign-up');
    const body = document.querySelector('body');
    expect(body).toMatchSnapshot();
  });

  it('debería imprimir la ruta home', () => {
    changeView('#/home');
    const body = document.querySelector('body');
    expect(body).toMatchSnapshot();
  });

  it('debería imprimir la ruta ""', () => {
    changeView('');
    const body = document.querySelector('body');
    expect(body).toMatchSnapshot();
  });

  it('debería imprimir ruta default', () => {
    changeView('#/otraruta');
    const body = document.querySelector('body');
    expect(body).toMatchSnapshot();
  });
});

describe('signupview', () => {
  const signUpView = createSignUpView();
  let email;
  let password;
  let btnSignUp;
  let eMessage;

  beforeAll(() => {
    document.body.innerHTML = '';
    document.body.appendChild(signUpView);
    email = document.getElementById('email');
    password = document.getElementById('userPassword');
    btnSignUp = document.getElementById('submitSingUp');
    eMessage = document.getElementById('eMessage');
    createBehaviorSignUpView();
  });

  it('debería devolver un elemento html', () => {
    expect(signUpView instanceof HTMLElement).toBe(true);
  });

  it('debería coincidir con el template', () => {
    expect(signUpView).toMatchSnapshot();
  });

  it('debería mostrar un modal cuando se registre correctamente', (done) => {
    email.value = 'prueba@hotmail.com';
    password.value = '123456';

    signUpWithEmail.mockResolvedValue({
      user: { email: 'prueba@hotmail.com', password: '123456', uid: 'fnvfyy' },
    });
    emailVerification.mockImplementationOnce(() => Promise.resolve());

    btnSignUp.click();

    setTimeout(() => {
      const modalContainer = document.querySelector('.modalContainer');
      expect(modalContainer.classList.contains('reveilModal')).toBe(true);
      done();
    }, 0);
  });

  it('Debería mostar mensaje de error si no tiene formato válido', () => {
    email.value = 'prueba';
    password.value = '123456';

    btnSignUp.click();
    expect(eMessage.textContent).toEqual('Debe ingresar un correo electrónico válido');
  });
});
