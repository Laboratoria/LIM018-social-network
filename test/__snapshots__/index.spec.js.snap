// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`changeView debería imprimir la ruta "" 1`] = `
<body>
  <div
    class="backgroundImage"
  >
    
  
    <div
      class="containerForm"
    >
      
    
      <div
        class="view"
      >
        
      
        <h1
          class="titles"
        >
           THE SOCIAL FOOD 
        </h1>
        
      
        <form>
          
        
          <input
            class="emailPassword"
            id="userEmail"
            placeholder="Email"
            type="email"
            value=""
          />
          
        
          <br />
          
        
          <input
            class="emailPassword"
            id="password"
            placeholder="Contraseña"
            type="password"
            value=""
          />
          
        
          <br />
          
        
          <button
            class="buttonsForm"
            id="submitLogIn"
            type="button"
          >
            Log In
          </button>
          
      
        </form>
        
      
        <div
          id="eMessage"
        />
        
        
        <p>
          o
        </p>
        
        
        <img
          class="googleImage"
          id="gmailLogIn"
          src="./images/google-logo.png"
        />
        
        
        <p>
          Con Google
        </p>
        
        
        <h3>
          Si no tienes cuenta, crea una 
          <span>
            <a
              href="#/sign-up"
            >
                aquí
            </a>
          </span>
        </h3>
        
    
      </div>
      
    
      <div
        class="modalContainer"
      >
        
      
        <div
          class="modal"
        >
          
        
          <img
            src="./images/email-alert-icon.png"
          />
          
        
          <p>
            Cuenta no verificada, porfavor revise su bandeja de correo electrónico
          </p>
          
        
          <button
            class="modalButton"
          >
            Aceptar
          </button>
          
      
        </div>
        
    
      </div>
          
  
    </div>
      
  </div>
</body>
`;

exports[`changeView debería imprimir la ruta log-in 1`] = `
<body>
  <div
    class="backgroundImage"
  >
    
  
    <div
      class="containerForm"
    >
      
    
      <div
        class="view"
      >
        
      
        <h1
          class="titles"
        >
           THE SOCIAL FOOD 
        </h1>
        
      
        <form>
          
        
          <input
            class="emailPassword"
            id="userEmail"
            placeholder="Email"
            type="email"
            value=""
          />
          
        
          <br />
          
        
          <input
            class="emailPassword"
            id="password"
            placeholder="Contraseña"
            type="password"
            value=""
          />
          
        
          <br />
          
        
          <button
            class="buttonsForm"
            id="submitLogIn"
            type="button"
          >
            Log In
          </button>
          
      
        </form>
        
      
        <div
          id="eMessage"
        />
        
        
        <p>
          o
        </p>
        
        
        <img
          class="googleImage"
          id="gmailLogIn"
          src="./images/google-logo.png"
        />
        
        
        <p>
          Con Google
        </p>
        
        
        <h3>
          Si no tienes cuenta, crea una 
          <span>
            <a
              href="#/sign-up"
            >
                aquí
            </a>
          </span>
        </h3>
        
    
      </div>
      
    
      <div
        class="modalContainer"
      >
        
      
        <div
          class="modal"
        >
          
        
          <img
            src="./images/email-alert-icon.png"
          />
          
        
          <p>
            Cuenta no verificada, porfavor revise su bandeja de correo electrónico
          </p>
          
        
          <button
            class="modalButton"
          >
            Aceptar
          </button>
          
      
        </div>
        
    
      </div>
          
  
    </div>
      
  </div>
</body>
`;

exports[`changeView debería imprimir la ruta sign-up 1`] = `
<body>
  <section
    class="backgroundImage"
  >
    
  
    <div
      class="containerForm"
    >
      
    
      <div
        class="view"
      >
        
      
        <h1
          class="titles"
        >
           THE SOCIAL FOOD 
        </h1>
        
        
        <form>
          
          
          <input
            class="registerInputs"
            id="userName"
            placeholder="Nombre de Usuario"
            type="text"
          />
          
          
          <br />
          
          
          <input
            class="registerInputs"
            id="email"
            placeholder="Correo electrónico"
            type="email"
            value=""
          />
          
          
          <br />
          
          
          <input
            class="registerInputs"
            id="userPassword"
            placeholder="Contraseña"
            type="password"
            value=""
          />
          
          
          <br />
          
          
          <button
            class="buttonsForm"
            id="submitSingUp"
            type="button"
            value="Sign Up"
          >
            Sign Up
          </button>
          
        
        </form>
        
        
        <div
          id="eMessage"
        />
        
        
        <img
          class="googleImage"
          id="gmailSignIn"
          src="./images/google-logo.png"
        />
        
        
        <p>
          Con Google
        </p>
        
        
        <h3>
          ¿Tienes cuenta?
          <span>
            <a
              href="#/log-in"
            >
                Entrar
            </a>
          </span>
        </h3>
        
    
      </div>
      
    
      <div
        class="modalContainer"
      >
        
      
        <div
          class="modal"
        >
          
        
          <img
            src="./images/verification-email.jpg"
          />
          
        
          <p>
            Se envió un código de verificación a su correo electrónico
          </p>
          
        
          <button
            class="modalButton"
          >
            Aceptar
          </button>
          
      
        </div>
        
    
      </div>
          
  
    </div>
  </section>
</body>
`;

exports[`changeView debería imprimir ruta default 1`] = `
<body>
  <div>
    
    
    <h2>
      404
    </h2>
    
    
    <h1>
      Página no encontrada
    </h1>
    
    
    <p>
      El archivo especificado no se encontró en este sitio web. Por favor, compruebe la URL para errores y vuelva a intentarlo.
    </p>
    
    
  </div>
</body>
`;

exports[`signupview debería coincidir con el template 1`] = `
<section
  class="backgroundImage"
>
  
  
  <div
    class="containerForm"
  >
    
    
    <div
      class="view"
    >
      
      
      <h1
        class="titles"
      >
         THE SOCIAL FOOD 
      </h1>
      
        
      <form>
        
          
        <input
          class="registerInputs"
          id="userName"
          placeholder="Nombre de Usuario"
          type="text"
        />
        
          
        <br />
        
          
        <input
          class="registerInputs"
          id="email"
          placeholder="Correo electrónico"
          type="email"
          value=""
        />
        
          
        <br />
        
          
        <input
          class="registerInputs"
          id="userPassword"
          placeholder="Contraseña"
          type="password"
          value=""
        />
        
          
        <br />
        
          
        <button
          class="buttonsForm"
          id="submitSingUp"
          type="button"
          value="Sign Up"
        >
          Sign Up
        </button>
        
        
      </form>
      
        
      <div
        id="eMessage"
      />
      
        
      <img
        class="googleImage"
        id="gmailSignIn"
        src="./images/google-logo.png"
      />
      
        
      <p>
        Con Google
      </p>
      
        
      <h3>
        ¿Tienes cuenta?
        <span>
          <a
            href="#/log-in"
          >
              Entrar
          </a>
        </span>
      </h3>
      
    
    </div>
    
    
    <div
      class="modalContainer"
    >
      
      
      <div
        class="modal"
      >
        
        
        <img
          src="./images/verification-email.jpg"
        />
        
        
        <p>
          Se envió un código de verificación a su correo electrónico
        </p>
        
        
        <button
          class="modalButton"
        >
          Aceptar
        </button>
        
      
      </div>
      
    
    </div>
        
  
  </div>
</section>
`;
